### DISCUSSION TITLE
Request to Develop or Contribute to an Elixir Client Library for Supabase

### DISCUSSION THREAD

**Discussion by bymayanksingh** [supabase/supabase#28343] [Contribute to Supabase]
## Chore

Hello Supabase Community ✨

I am interested in contributing to the Supabase project by developing a client library for Elixir. 

Before proceeding, I would like to check if there is already an existing Elixir client library for Supabase. 

If such a library exists, I would be happy to contribute to its development by collaborating with the author and improve it. 

If not, I would like to propose the creation of a new Elixir client library from scratch.

## Describe the chore

This will help in expanding Supabase's ecosystem with support for Elixir, given few projects are built using Elixir itself within Supabase, I believe this will be a great addition.

This will also help in engaging with the Elixir community and attracting more developers to use Supabase.

We can povide Elixir developers with a robust and well-maintained client library to interact with Supabase services.

Once the proposal is accepted, I welcome any input, suggestions, or collaboration from fellow developers who are interested in working on this project

Thank you.

PS: let me know, if this is not the right repo to open this issue, I'll re raise the chore in another respective repo.

cc @awalias @kiwicopple

**Comment by encima** [comment]
Thanks for asking @bymayanksingh! 

There is nothing officially supported (or within Supabase community) for Elixir as far as I know, though there is [Supabase-Ex](https://github.com/zoedsoupe/supabase-ex)

I moved this to a discussion for contributing and will let the @supabase/devrel respond to the actual request.

**Reply by bymayanksingh** [reply]
Sure, thank you for opening a discussion around this.

**Reply by bymayanksingh** [reply]
Who from the Developer Relations team can we tag here @encima ?

**Reply by encima** [reply]
Spoke with the team and the existing library is a good starting point. @zoedsoupe would be the person to reach out to so I would propose this on the `supabase-ex` repo as an offer to contribute or develop a library that covers missing features.
From there, we can add it to `supabase-community`

Let us know the outcome (here or you can email) and we can help moving forward.

Thanks again for the offer of help!

**Reply by zoedsoupe** [reply]
Hi @bymayanksingh @encima,

I appreciate your enthusiasm for contributing to the Supabase ecosystem! I wanted to share that we already have a complete feature client library for Elixir called [supabase-ex](https://github.com/zoedsoupe/supabase-ex). This library has been in intensive development and includes support for the following Supabase services:

- [Supabase PostgREST](https://github.com/zoedsoupe/postgrest-ex)
- [Supabase GoTrue](https://github.com/zoedsoupe/gotrue-ex)
- [Supabase Storage](https://github.com/zoedsoupe/storage-ex)
- [Supabase-Ex (the main project)](https://github.com/zoedsoupe/supabase-ex)

And the respective https://hex.pm packages:

- https://hex.pm/packages/supabase_potion
- https://hex.pm/packages/supabase_gotrue
- https://hex.pm/packages/supabase_storage
- https://hex.pm/packages/supabase_postgrest

I’ve presented this project to the Supa Squad since last year, and it has been actively worked on ever since. While the library does need some refactoring, it already provides robust support for interacting with Supabase services.

You can find the [last year spa squad presentation](https://github.com/zoedsoupe/supabase-ex/blob/main/supaquad_presentation_08_11_2023/presentation.md) and [this year presentation too](https://github.com/zoedsoupe/supabase-ex/blob/main/supaquad_presentation_15_05_2024/presentation.md)

Also notice that supabase community already "suggest" some old legacy and non maintained libraries that you can find on https://github.com/supabase-community/postgrest-ex and https://github.com/supabase-community/gotrue-ex. My effort was to refactor and cover more feature than these legacy libraries.

Given that `supabase-ex` is already well-established, it would be fantastic to have contributions and refactoring efforts to improve it further, rather than starting a new library from scratch. This way, we can pool our efforts to make `supabase-ex` even better and more comprehensive.

If you’re interested, I’d love to collaborate and welcome any contributions you can make to the library. Let’s work together to enhance the Elixir client library for Supabase!

Feel free to reach out to me, and we can discuss the next steps.

Thanks,
@zoedsoupe [zoey.spessanha@zeetech.io](mailto:zoey.spessanha@zeetech.io)

**Reply by bymayanksingh** [reply]
Hi @zoedsoupe,

Thank you so much for the detailed response and for outlining the current state of [supabase-ex](https://github.com/zoedsoupe/supabase-ex). I'm excited to hear that there's already a comprehensive Elixir client library in place!

I would love to contribute to supabase-ex and help take it to the next level, ensuring it becomes a robust and well-maintained tool for the Supabase/Elixir community. I'm particularly interested in collaborating on the necessary refactoring and any new features that can enhance its functionality.

Thank you, @encima, for initiating this discussion and guiding it in the right direction. Your support is greatly appreciated.

Looking forward to working together on this one @zoedsoupe @encima, to make [supabase-ex](https://github.com/zoedsoupe/supabase-ex) an even greater success!

Let me where we can discuss the next steps @zoedsoupe @encima.

Best,
Mayank Singh

**Reply by zoedsoupe** [reply]
@bymayanksingh sorry for the delayed response, i got very busy on these last weeks, but i have some good news! I just launched new updated and refactored versions for the aforementioned hex packages, targeting the base SDK (aka `supabase-ex`) implementation.

to summarize, i simplified the client initialization and move the decision to manage OTP processes for the final user, as it should be! this should also prevent strange errors on starting and using clients but also clarify the documentation on **how** to start a client, as in the later version it had around 3 ways to start a client...

so basically, the current state for each library and my current roadmap is (very summarized):

## [supabase-ex](https://github.com/zoedsoupe/supabase-ex)
- easy manipulation, configuration and initialization of clients
  - supports all available configuration options as `supabase-js`
  - provides a pre defined "high-level" http client for supabase APIs (it should be refactored in some points)

> it have a total of 2422 downloads on hex

## [postgrest-ex](https://github.com/zoedsoupe/postgrest-ex)
- follows the standard of `postgrest-js` and `postgrest-go` with composable and "lazy" query and filter builders
- have good error mapping structure
- can not only execute queries and return JSON response but also can parse to an Ecto schema/struct
- good test coverage

> it have a total of 242 downloads on hex
> elixir community will very often prefer to connect to the postgresql instance directly with ecto

## [storage-ex](https://github.com/zoedsoupe/storage-ex)
- handle all bucket operations
- handle all object operations
  - with addition of stream upload
  - with addition of stream download

> it have a total of 802 downloads on hex
> for some time the package was broken on hex 😕

## [gotrue-ex](https://github.com/zoedsoupe/gotrue-ex)
- it have [Plug](https://hexdocs.pm/plug) applications support (plug middlewares)
- it have [Phoenix.LiveView](https://hexdocs.pm/phoenix_live_view) applications support (hooks)
- implements password reset email
- implements resend of confirmation email
- implements user update (also admin API)
- implements signup flow
- implements user creation (admin)
- implements signing in with
  - id token
  - oauth
  - otp
  - email and password 
  - sso
- implements sign out (admin)
- implements email invitation and link generation (eg. any verification type) (admin)
- implements user management (deletion, list, get by id, update by id) (admin)

> it have a total of 1738 downloads on hex

## Roadmap

Said that, i would like to tackle these next features:
- Supabase.UI library for `Phoenix.LiveView`
- Supabase Realtime library and integration
  - as the realtime project is also in Elixir, i was discussing with @filipecabaco about the possibility to create adapters for `PubSub` and or process messaging to receive updates, besides the API integration
- Implements anonymous sign in on `gotrue-ex`
- Improvements and refactors on the PKCE flow on `gotrue-ex`
- integration with  Ecto  via custom adapter to be able to use  Ecto.Query
  easily on `postgrest-ex`
  - also consider [this issue/feature request](https://github.com/zoedsoupe/supabase-ex/issues/4#issuecomment-2113315471)

**Reply by zoedsoupe** [reply]
@bymayanksingh any updates? also i know the old work from @treebee and @J0 were maintainers. does you folks intend to continue giving support? i kinda already did a complete rewrite and i'm now fixing some bugs and focusing on documentation

**Reply by bymayanksingh** [reply]
Hey, really sorry, I won't be able to continue on this as of now, I am now committed to few other demanding projects & it's getting a little hard to manage all at once.

**Reply by zoedsoupe** [reply]
no problem on that! very relatable indeed
